### PR TITLE
Removes NTSL

### LINKED ISSUE #1974: Telecomms traffic computer can be set to run scripts on the Syndicate frequency
##### Server Revision:
12/4/2013

##### Server:
Probably Sibyl and Basil although I'm not so sure since this is a secondhand issue.

###### What did you expect to happen? 
The Syndicate frequency is secure from NTSL tampering and thus can't be eavesdropped upon.

###### What happened instead?
Budding Engineers can set that frequency to be sniffed by receivers/etc and can thus run NTSL upon the frequency.

###### Why is this bad/What are the consequences?
Engineers can eavesdrop upon it to instantly know if it's a nuke round or not.

##### Steps to reproduce the problem:
1. Multitool the Telecomms equipment 
2. Set the receiver/etc to treat the Nuke Ops freq exactly like 149.5.
3. Set up a script to run on the common server that if $freq = nuke freq to send alerts or whatever.

##### Possibly related stuff:
Can also do this with the Deathsquad frequency

##### Anything else?

### LINKED ISSUE #4710: NTSL/Saycode: Hearing radio message requires hearing other radio message
**Server Revision:** c487dd350a666c73f22abc296182f05863f1186d

**Which server were you playing on?**
Sibyl
Also tested locally to narrow the problem

**Problem Description**:
As an AI, I'm used to using NTSL message frequency redirections (modifying $freq) to keep names clickable. However, hearing messages through one intercom now sometimes requires hearing another message through another intercom.

**Steps to reproduce the problem**:
1. Be roundstart AI
2. Upload and enable NTSL script:
```
broadcast($content, $freq, $source, $job);
$freq = 1447;
```
3. say "Test"
4. Observe being able to hear both messages through both intercoms
5. Turn common intercom speaker off
6. say "Test"
7. Observe not being able to hear anything through either intercom, including the one that's still on

**Anything else?**
If the clown honks on a station and saycode doesn't let you hear it, does he truly make a sound?

### LINKED ISSUE #5779: Telecoms script parser runtime
```
runtime error: Cannot read null.value
proc name: CheckToken (/n_Parser/nS_Parser/proc/CheckToken)
source file: Parser.dm,131
usr: Purity (/mob/living/silicon/robot)
src: /n_Parser/nS_Parser (/n_Parser/nS_Parser)
call stack:
/n_Parser/nS_Parser (/n_Parser/nS_Parser): CheckToken(")", /token/symbol (/token/symbol), 1, 1)
/n_Keyword/nS_Keyword/kwIf (/n_Keyword/nS_Keyword/kwIf): Parse(/n_Parser/nS_Parser (/n_Parser/nS_Parser))
/n_Parser/nS_Parser (/n_Parser/nS_Parser): Parse()
/datum/TCS_Compiler (/datum/TCS_Compiler): Compile("if($job==\"AI) {$job=\"Dungeon...")
Engineering Server (/obj/machinery/telecomms/server/presets/engineering): compile(Purity (/mob/living/silicon/robot))
```

### LINKED ISSUE #10032: AI tracking is broken
After the new AI tracking update that creates a delay on AI tracking I've played two rounds as the AI where I would try to track people from the chat only to have them be untrackable.  They would be near cameras once they told me where they were and I could be looking right at them, but they were still untraceable.  

This happened with a good chunk of the crew, including sec officers who shouldn't have been given syndicate antag roles which might have given them access to syndicate ID's

The only thing I noticed that seemed to be consistent with this was the problems seemed to start after people started making changes to t-com scripts, usually to add job info to people  when they spoke.

### LINKED ISSUE #10445: it's very easy to lock up the server with NTSL scripts
I won't provide the code here for obvious reasons, but it involves lists and gets around the instruction limit.

### LINKED ISSUE #10595: NTSL Only renders the first few characters
My god how old is this bug?

### LINKED ISSUE #11647: Remove NTSL
Or fix it, I guess, but it can do bad stuff to the server right now.
Details will not be posted if you've been in coderbus recently, you know what's wrong and how bad it is.

### PR DESCRIPTION
**Pull request by MrPerson**
Reasoning for why it needs to go:
- NTSL allows anyone to lag the server or crash it.
- Doohl being gone, no one is going to maintain or expand NTSL. Very little has been done on it for several years now, certainly not any new functionality besides signalling.
- The only interesting things that can be done with NTSL have already been done and are on the wiki. Which is what 99% of users do, copy a script off the wiki. Not interesting or fun.
- Users who actually do scripting are too limited to actually do much besides little chat scripts that are amusing for maybe 2 minutes before becoming annoying.

Fixes #11647
Fixes #10595
Fixes #10445
Technically fixes #10032
Fixes #5779
Technically fixes #4710
Fixes #1974

### CODE REVIEW

**Comment by Lumipharon** [comment]
Will miss some neat things, like job tags and muting individual people as a rogue AI, but yeah, fuck the lag this shit causes.

**Comment by Bawhoppen** [comment]
Mixed feelings, NTSL is horrible but its fun at the same time. Would be sad to see it go.

**Comment by WJohn** [comment]
Thank you so much.

**Comment by Giacom** [comment]
I think NTSL was kinda pushing the limits of BYOND. It only ran when someone talked, and was generally not slow as everything was parsed into objects, so it did not create constant slowdown but I suppose if there was a lot of chatter then it could have affected the server.

I can also see how this could be a bother to maintain.

**Comment by Dorsisdwarf** [comment]
RIP brave soldier go in peace

**Comment by Supermichael777** [comment]
have you ensured tcoms will run without this? (goof has taught me never to assume anything)

also we seem to be (trying) axing a ton of features lately.
could you add the ability for malf ais to selectively mute people after this?

File: _maps/map_files/MetaStation/MetaStation.v41G.dmm [hunk 1]
```diff
@@ -4017,7 +4017,7 @@
 "bzm" = (/obj/machinery/atmospherics/pipe/manifold/supply/hidden{dir = 8},/turf/simulated/wall/r_wall,/area/atmos)
 "bzn" = (/obj/machinery/atmospherics/pipe/simple/supply/hidden{dir = 4},/turf/simulated/wall/r_wall,/area/atmos)
 "bzo" = (/obj/machinery/atmospherics/pipe/simple/supply/hidden{dir = 10},/turf/simulated/wall/r_wall,/area/atmos)
-"bzp" = (/obj/machinery/computer/telecomms/traffic{network = "tcommsat"},/turf/simulated/floor/plasteel{icon_state = "grimy"},/area/tcommsat/computer{name = "\improper Telecoms Control Room"})
+"bzp" = (/obj/structure/table/wood/poker,/obj/effect/spawner/lootdrop{loot = list(/obj/item/weapon/gun/projectile/revolver/russian = 5, /obj/item/weapon/storage/box/throwing_stars, /obj/item/toy/cards/deck/syndicate = 2); name = "gambling valuables spawner"},/turf/simulated/floor/wood,/area/crew_quarters/bar)
```
Comment (Supermichael777) [inline _maps/map_files/MetaStation/MetaStation.v41G.dmm]: throwing stars? what are you trying to pull

Comment (Dorsisdwarf) [inline _maps/map_files/MetaStation/MetaStation.v41G.dmm]: He's added a poker table with a lootdrop including the (possibility of) russian revolvers, syndiecards, and throwing stars where the main console used to be.

**Comment by Dorsisdwarf** [comment]
I will definitely miss the ability to make all humans hiss and all lizards speak normally, and the clown speak backwards.

File: _maps/map_files/MetaStation/MetaStation.v41G.dmm (hunk 1 above)
Comment (MrPerson) [inline _maps/map_files/MetaStation/MetaStation.v41G.dmm]: Blame it on the map merger. It's merging another identical tile of exactly this together while leaving the tile of the traffic computer as a different duplicate.

**Comment by GunHog** [comment]
:-1: x 100
Am I the only one here who discovered the fun and useful things one can do with NTSL? Even all these years it remains the most flexible feature in the whole game. Scripting with NTSL is what inspired me to code. The exploit can be fixed.

Those little wiki scripts are a pale comparison to what can be done. It is like only doing arithmetic with a graphing calculator. I am surprised so few players know how amazing it truly can be. I cannot fix a player's lack of creativity - that is a problem widespread in our community.

**Comment by Durkel** [comment]
With the fact it is causes the server to occasionally shit itself this was inevitable. Goodbye ntsl, you won't be missed that much.

**Comment by Dorsisdwarf** [comment]
@GunHog 

Yes, they did.

They mass-copied your script.

**Comment by Cheridan** [comment]
>Am I the only one here who discovered the fun and useful things one can do with NTSL? 

yes

**Comment by RemieRichards** [comment]
Ultimately, this is for the best.
if ByondTCL ever becomes a thing, we can add that later.

**Comment by GunHog** [comment]
MSO may be working on the exploit fix, that is your only real motivation to remove it besides simply not liking it.

**Comment by Giacom** [comment]
What's ByondTCL?

**Comment by RemieRichards** [comment]
@Giacom only an experiment atm, and MSO stopped working on it for a bit because byond is ass with strings, but it's an implementation of a smaller (only 12 syntax rules) scripting language (TCL, Tool Command Language) in Byond. the reason strings are so important to TCL is that in TCL EVERYTHING, including functions and language syntax, is a string.

The idea was it would fill NTSL's place, without being a sprawling hardly understood psuedo language.

**Comment by Razharas** [comment]
It is a feature that takes shitloads of code, is exceptionally complex and can be cpu heavy, yet its usefulness is miniscule
If we at least had some programmable bots to use this in maybe it would see some use, but currently only CE and AI have access to this, it affects only radio chat and is too bothersome to really do anything new with
In the current state ntsl is not worth the code it is occupying, if you feel enthusiastic you can try to renew the ntsl with something that will fix the abovementioned problems @GunHog 
In that case we will be deciding between removing and improving, as of now removing is the only option

**Comment by GunHog** [comment]
TCL could take a year or more to get added, and you are removing this now...I can try coding something to work with bots, although that can get weird now that radio code is gone from them. Probably name matching or something.

Right now it only works on Tcomms, yes, but without it, there is no way to do things such as changing your radio name on the fly, displaying jobs, giving people custom voice fonts, etc.

I am not trained in the art of debate, so I cannot convince people easily. It is obvious that you are going to remove NTSL regardless of what I say, but could you at least replace it with the features I add to the radio from my script? I have found NTSL to be one of the best features in the game, and it is a shame you are removing it.

**Comment by Anonus** [comment]
NTSL is a nice thing to have but it can crash the server without any warning, and having a largely easily accessable way to do that is, uh, pretty bad.

**Comment by GrayRachnid** [comment]
R I P

It'll be nice to have some kind of alternative for muting people as the AI/Engineer or deceiving people in general. And job titles.

**Comment by SconesC** [comment]
Watching MAGNI copypasted every round for obscene AI telecomms powergaming got old 

I wish job titles were standardized into the radio but otherwise this is :+1:

**Comment by Strathcona** [comment]
Will there be hardcoded features to replace the numerous telecoms features it provided? Things like the automated announcements system? We should probably replace at least some of the functionality, at the very least the useful things like job tags or the like

**Comment by Giacom** [comment]
Huh, is it similar to Tobba's door control system? Sounds cool.

For me I always wanted to have all the computers running on a stack based language or something with command line interfaces.

**Comment by RemieRichards** [comment]
@Giacom Wouldn't know, the only project of Tobba's I've seen is his 509 lighting.
Yeah I saw your stack based language, since you put it up on Github (Unless you have 2 lol)

**Comment by Giacom** [comment]
Nah it's just the 1. Thinking about it though, I would prefer having a "high level language" like BASIC.

* Doing things in a stack based language are slow since every operation is going to be painful to the server and you would need a lot.
* BASIC can stop and start very easily, it will likely not need many operations to do.
* It's less intimidating than NTSL or a stack based language, it might get more people interested.

**Comment by RemieRichards** [comment]
Well I assume that's the same reasons TCL intrigued MSO, it's 12 syntax rules, Simple to implement and then simple for players to use.

**Comment by Giacom** [comment]
Let me know if anything is written down about that other than being just an idea in a head; I am interested in taking a look.

**Comment by RemieRichards** [comment]
@Giacom https://github.com/tcl4dm/tcl4dm You could get started if you -really- wanted. Last thing me and MSO discussed was using a custom string datum with a list of text2ascii, to avoid byond's strings as much as possible.

**Comment by Giacom** [comment]
Hmm, last modification was made 4 months ago, that's a little worrying.

With the custom string datums, my initial guess would be that the native stuff in BYOND would be faster than hand made datums. Is that not the case for BYOND strings?

**Comment by Jalleo** [comment]
@giacom go and talk to mso about it via the forums or irc there are ideas of how to implement it and @Mrstonedone did mention something about another more complex way with datums that could work faster since byond is unoptimised with strings.

**Comment by optimumtact** [comment]
:-1: I am against this in principle

**Comment by Fox-McCloud** [comment]
Are you referring to Goon's computer+radio+door system? That's a fairly interesting system, but it's fairly obscure, as well---it's also several orders more powerful than NTSL since you can do just about anything you want with it, once you understand it.

**Comment by MrPerson** [comment]
I went ingame and talked on public radio and specific frequencies and there were no runtimes.

**Comment by tuypo1** [comment]
if we do remove it (although I think it should stay) we should at least make job titles standard or at least put a button to enable it somewhere

**Comment by CPTANT** [comment]
I will miss dumping the captain in space and impersonating his voice on the radio as AI.

Perhaps a simpler ntsl system could be made more focussed on the end result, instead of being its own scripting language. Make predefined functions for muting/replacing words/replacing sources etc. That is basically how people already use it, by taking already made chunks of code and copy pasting it with or without some minor editing.

**Comment by Bawhoppen** [comment]
https://www.youtube.com/watch?v=_1qXQRpF08E

**Comment by AnturK** [comment]
Not that i disagree with this but can we stop with randomly instamerging shit.

**Comment by CPTANT** [comment]
^ that.

This is a significant removal and the discussion was no where near finished.

**Comment by GunHog** [comment]
Nor was it necessary because MSO was <b>working on a fix to this.</b>

**Comment by duncathan** [comment]
Yolomerging is a way for maintainers to agree on something while letting the blame only fall on one
Razh and Aran are good scapegoats
#RevealingTheCodespiracy
:^)

**Comment by Qbopper** [comment]
I have very strong feelings about this but I can't decide if I'm against or for it

I think some of the things you can do with it are very useful/neat as shit, but like you say it's laggy and shit

**Comment by Miauw62** [comment]
Why do we always have to quickmerge inelegant stuff.
Op 3-okt.-2015 23:27 schreef "Qbopper" <notifications@github.com>:

**Comment by tuypo1** [comment]
what AnturK said

**Comment by astralenigma** [comment]
Honestly I was working unsucsessfully on a script that would sound like a radio station, in one of the first drafts it ran so many times in a second that the game stopped the script after a good amount of operations if I understood correctly the message, I'm sad that NTSL is being removed but I understand but hope for a better implementation in the future.

**Comment by C-F-K** [comment]
God fucking dammit

I stop playing SS13 for a while to write my grad thesis and this is what happens?! 

Welp I'm disconsolate now. I made this before I stopped playing but it looks like I'm never going to get to use it so here it is for those of you inclined to appreciate this bullshit: [a telescience calculator](https://gist.github.com/C-F-K/e845c7581b128c3fb300)

**Comment by optimumtact** [comment]
Holy necro batman

**Comment by duncathan** [comment]
that's actually pretty impressive, not gonna lie.

**Comment by GunHog** [comment]
@C-F-K There are codebases (Hippie, Paradise) that still appreciate the unrivaled fun this feature used to provide, so do not give up hope! You can still use that extremely impressive piece of work!

**Comment by Giacom** [comment]
I would still like SS13 to have some kind of scripting gameplay mechanic. It fits the theme of deep mechanics.

**Comment by RemieRichards** [comment]
@Giacom probably won't happen unless MSO makes a good implementation of TCL, which granted would still have the same problems as NTSL:
* nobody uses it
* lag (some what mitigated by CHECK_TICK)

**Comment by CPTANT** [comment]
The real problem is that it completely relies on out of game effort.

The amount of time required to actually write a script in game basically forces you to copy and paste, which is rather lame.

**Comment by C-F-K** [comment]
@duncathan Thanks :) 

@GunHog Also thanks :) Failing that, I might try to finagle it back into my local repo and enjoy it on my own server at least.

@RemieRichards Shit, given enough time (ie. once I get my thesis done) I might dig into the code and learn to implement that myself. Or something similar. Kind of depends on what's going on with the codebase by that time.

**Comment by RemieRichards** [comment]
@C-F-K feel free too, I'd despite my apparent negativity I really would like a (stable) scripting system, as Gia says it really does fit SS13.

**Comment by GunHog** [comment]
I do not have faith in this codebase in terms of this. It is fun, powerful, and extremely modular. Such a thing is unlikely to be allowed again here, at least under our current leadership.

That being said, if you can pull it off, I will love and serve you forever.

**Comment by PKPenguin321** [comment]
Doesn't fix the fact that it's an almost entirely ooc process (not unlike telescience)

**Comment by Miauw62** [comment]
gunhog ntsl was literally only ever used by AIs to copypaste in the same script every fucking round

that's all it was ever used for. if somebody else changed it the ai would just copypaste its script over it again.

it wasnt fun or engaging

**Comment by Dorsisdwarf** [comment]
I agree that the silicons should have to ask/persuade/blackmail an under(over?)ling into doing scripts for them, if they ever come back

**Comment by C-F-K** [comment]
Telescience doesn't have to be OOC if you can do the math, which is pretty much why I wrote the script, actually &ndash; it stands to reason that if a scientist (and especially an experimental scientist &ndash; the jury's still out for theoretical ones, but we don't seem to have many of those on the station anyway) wasn't versed in some branch of math, they'd think "that's what computers are for" before they'd think "let me just learn that real quick". Given that, I'd also imagine that programming skill would be more common among scientists than trigonometry. 

:+1: for the idea of restricting the AI from accessing the comms server. I'm not sure how exactly silicon communications intersect with the rest of the comms system but it seems that that would make for some interesting gameplay with regard to resisting rogue AIs and such. (And come on, someone in the engineering department has to be the server room nerd.)

**Comment by CPTANT** [comment]
Doing the math is exactly what the OOC part is >_>

**Comment by C-F-K** [comment]
Did you read what I said tho :v It's less OOC for a scientist to bash out my NTSL script (read: copy-paste, their player making the assumption that their character would know how to write it) than it would be for the same player to just use an out-of-game calculator. 

Still a little bit OOC, I agree, but the need for something as OP as telescience to be limited somehow (which function is served by the math) needs to be balanced against the need for the game to be essentially accessible to everyone (including people who don't have the time and energy to learn trig). Combined with the fact that the telesci lab is no longer built by default (serving as an additional limiting factor), I think my script is a good compromise between the two.

**Pull request merged**